File: PALIN/palin.hs
import Control.Arrow ((>>>), first)

import Data.Word (Word8)

import Data.ByteString (ByteString)
import qualified Data.ByteString as BS
import qualified Data.ByteString.Char8 as C8


add :: Word8 -> (Word8, Bool)
add w 
    | 48 <= w && w < 57 = (w + 1, False)
    | w == 57 = (48, True)
    | otherwise = error $ show w


addOne :: ByteString -> (ByteString, Bool)
addOne =
    BS.foldr addOne' ([], True)
    >>> (\(bs, c) -> if c then (49:bs, c) else (bs, c))
    >>> first BS.pack
  where
    addOne' word (bs, c)
        | c         = let (word', c') = add word in (word': bs, c')
        | otherwise = (word : bs, c)


leqStr :: ByteString -> ByteString -> Bool
leqStr str1 = leqStr' 0 (BS.length str1) str1
  where
    leqStr' i n str1 str2
        | i >= n = True
        | BS.index str1 (n-i-1) < BS.index str2 i = True
        | BS.index str1 (n-i-1) > BS.index str2 i = False
        | otherwise = leqStr' (i+1) n str1 str2


transform :: ByteString -> ByteString
transform bs = BS.append left'' (BS.reverse right')
  where
    len = BS.length bs
    e = even len
    h = div len 2 + (if e then 0 else 1)
    r = if e then h else h - 1
    left = BS.take h bs
    right = BS.drop r bs
    (left', carry)
        | leqStr left right = addOne left
        | otherwise = (left, False)
    right' = if e then left' else BS.init left'
    left'' = if carry then BS.init left' else left'


main :: IO ()
main = do
    t <- read <$> getLine
    mainLoop t


mainLoop :: Int -> IO ()
mainLoop 0 = pure ()
mainLoop t = do
    BS.getLine >>= (BS.split 13 >>> head >>> transform >>> C8.putStrLn)
    mainLoop (t-1)

File: PALIN/palin.py
import sys

ADD = [
    ('1', False),
    ('2', False),
    ('3', False),
    ('4', False),
    ('5', False),
    ('6', False),
    ('7', False),
    ('8', False),
    ('9', False),
    ('0', True),
]

def add_one(string):
    n = len(string)
    to_return = ''
    
    digit, carry = ADD[ord(string[-1])-48]
    to_return = digit+to_return
    for i in range(2, n+1):
        if carry:
            digit, carry = ADD[ord(string[-i])-48]
        else:
            digit = string[-i]
        to_return = digit + to_return
    if carry:
        to_return = '1' + to_return
    return to_return, carry


def leq_str(string1, string2):
    for i, c in enumerate(string2):
        if string1[-(i+1)] < c:
            return True
        elif string1[-(i+1)] > c:
            return False
    return True

_ = int(input())


for line_raw in sys.stdin:
    # half line
    line, = line_raw.split()
    l = len(line)
    even = l % 2 == 0
    h = l // 2 + (0 if even else 1)

    r = h if even else h-1
    carry = False
    if leq_str(line[:h], line[r:]):
        left, carry = add_one(line[:h])
    else:
        left = line[:h]

    right = left if even else left[:-1]
    left = left[:-1] if carry else left
    print(left+right[::-1])